### PR TITLE
Throttle redis custom keys

### LINKED ISSUE #302: Feature: Add custom limiter keys and parsing limits from specified json field in redis value
Current implementation of redis limiters in throttle plugin allows only default limiter keys formed using pipeline name, throttle field and throttle field value, and getting limits as direct values under limiter keys in redis.
It would be great if redis limiter supported custom limiter keys and parsing limits from at least first level of json object.

### PR DESCRIPTION
**Pull request by HeadHunter483**
Fixes #302

### CODE REVIEW

File: plugin/action/throttle/throttle.go [hunk 1]
```diff
@@ -338,6 +354,13 @@ func (p *Plugin) isAllowed(event *pipeline.Event) bool {
 			throttleKey = val
 		}
 	}
+	keyLimitOverride := ""
+	if len(p.config.RedisBackendCfg.LimiterKeyField_) > 0 {
+		val := event.Root.Dig(p.config.RedisBackendCfg.LimiterKeyField_...).AsString()
+		if val != "" {
+			keyLimitOverride = val
+		}
+	}
```
Comment (DSmolonogov) [inline plugin/action/throttle/throttle.go:363]: may be `keyLimitOverride := event.Root.Dig(p.config.RedisBackendCfg.LimiterKeyField_...).AsString()`

**Review by DSmolonogov — APPROVED** [review]

File: plugin/action/throttle/throttle.go (hunk 1 above)
Comment (HeadHunter483) [reply plugin/action/throttle/throttle.go]: Done

File: plugin/action/throttle/redis_limiter.go [hunk 2]
```diff
@@ -180,14 +194,62 @@ func (l *redisLimiter) updateLimiterValues(maxID, bucketIdx int, totalLimiterVal
...
+}
+
 // updateKeyLimit reads key limit from redis and updates current limit.
-func (l *redisLimiter) updateKeyLimit() {
+func (l *redisLimiter) updateKeyLimit() error {
 	// try to set global limit to default
-	if b, err := l.redis.SetNX(l.keyLimit, l.totalLimiter.limit.value, 0).Result(); err == nil && !b {
+	val := strconv.FormatInt(l.totalLimiter.limit.value, 10)
+	if l.valField != "" {
+		val = fmt.Sprintf(`{"%s":"%s"}`, l.valField, val)
```
Comment (vadimalekseev) [inline plugin/action/throttle/redis_limiter.go:227]: Do we need escape `limiter_value_field`?
```suggestion
	key, _ := json.Marshal(l.valField) // todo handle error, do it once in the `.Start`
	val = fmt.Sprintf(`{"%s":"%s"}`, key, val)
```

File: plugin/action/throttle/redis_limiter.go [hunk 3]
```diff
@@ -66,7 +74,11 @@ func NewRedisLimiter(
 	rl.keyPrefix.WriteString("_")
 	rl.keyPrefix.WriteString(throttleFieldValue)
 	rl.keyPrefix.WriteString("_")
-	rl.keyLimit = rl.keyPrefix.String() + keySuffix
+	if keyLimitOverride == "" {
```
Comment (vadimalekseev) [inline plugin/action/throttle/redis_limiter.go:77]:
```suggestion
	if keyLimitOverride == "" {
		rl.keyPrefix = bytes.Buffer{}
		// full name of keyPrefix will be pipelineName_throttleFieldName_throttleFieldValue_limit. `limit` added afterwards
		rl.keyPrefix.WriteString(pipelineName)
		rl.keyPrefix.WriteString("_")
		rl.keyPrefix.WriteString(throttleFieldName)
		rl.keyPrefix.WriteString("_")
		rl.keyPrefix.WriteString(throttleFieldValue)
		rl.keyPrefix.WriteString("_")
		rl.keyLimit = rl.keyPrefix.String() + keySuffix
	} else {
		rl.keyLimit = keyLimitOverride
	}
```

File: plugin/action/throttle/throttle.go [hunk 4]
```diff
@@ -338,6 +354,10 @@ func (p *Plugin) isAllowed(event *pipeline.Event) bool {
 			throttleKey = val
 		}
 	}
+	keyLimitOverride := ""
+	if len(p.config.RedisBackendCfg.LimiterKeyField_) > 0 {
+		keyLimitOverride = event.Root.Dig(p.config.RedisBackendCfg.LimiterKeyField_...).AsString()
```
Comment (vadimalekseev) [inline plugin/action/throttle/throttle.go:359]: Do we need clone the string?

```suggestion
		keyLimitOverride = strings.Clone(event.Root.Dig(p.config.RedisBackendCfg.LimiterKeyField_...).AsString())
```

File: plugin/action/throttle/redis_limiter.go [hunk 5]
```diff
@@ -180,14 +194,62 @@ func (l *redisLimiter) updateLimiterValues(maxID, bucketIdx int, totalLimiterVal
 	l.totalLimiter.mu.Unlock()
 }
 
+func getLimitValFromJson(data []byte, valField string) (int64, error) {
```
Comment (vadimalekseev) [inline plugin/action/throttle/redis_limiter.go:197]:
```suggestion
func getLimitValFromJson(data []byte, valField string) (int64, error) {
	var m map[string]json.Number
	if err := json.NewDecoder(bytes.NewReader(data)).Decode(&m); err != nil {
		return 0, fmt.Errorf("failed to unmarshal map: %w", err)
	}
	limitVal, has := m[valField]
	if !has {
		return 0, fmt.Errorf("no %q key in map", valField)
	}
	return limitVal.Int64()
}
```

File: plugin/action/throttle/redis_limiter.go [hunk 6]
```diff
@@ -180,14 +194,62 @@ func (l *redisLimiter) updateLimiterValues(maxID, bucketIdx int, totalLimiterVal
...
+	val := strconv.FormatInt(l.totalLimiter.limit.value, 10)
+	if l.valField != "" {
+		val = fmt.Sprintf(`{"%s":"%s"}`, l.valField, val)
+	}
+	if b, err := l.redis.SetNX(l.keyLimit, val, 0).Result(); err == nil && !b {
 		// global limit already exists - overwrite local limit
-		if v, err := l.redis.Get(l.keyLimit).Int64(); err == nil {
-			l.totalLimiter.limit.value = v
-			l.incrementLimiter.limit.value = v
+		if v := l.redis.Get(l.keyLimit); err == nil {
```
[the remark this answers is not in the document]
Comment (HeadHunter483) [reply plugin/action/throttle/redis_limiter.go]: done

File: plugin/action/throttle/redis_limiter.go [hunk 7]
```diff
@@ -180,14 +194,62 @@ func (l *redisLimiter) updateLimiterValues(maxID, bucketIdx int, totalLimiterVal
...
 // updateKeyLimit reads key limit from redis and updates current limit.
-func (l *redisLimiter) updateKeyLimit() {
+func (l *redisLimiter) updateKeyLimit() error {
 	// try to set global limit to default
-	if b, err := l.redis.SetNX(l.keyLimit, l.totalLimiter.limit.value, 0).Result(); err == nil && !b {
+	val := strconv.FormatInt(l.totalLimiter.limit.value, 10)
+	if l.valField != "" {
+		val = fmt.Sprintf(`{"%s":"%s"}`, l.valField, val)
+	}
+	if b, err := l.redis.SetNX(l.keyLimit, val, 0).Result(); err == nil && !b {
```
[the remark this answers is not in the document]
Comment (HeadHunter483) [reply plugin/action/throttle/redis_limiter.go]: done

File: plugin/action/throttle/redis_limiter.go (hunk 2 above)
Comment (HeadHunter483) [reply plugin/action/throttle/redis_limiter.go]: done, moved to new limiter creation func

File: plugin/action/throttle/throttle.go (hunk 4 above)
Comment (HeadHunter483) [reply plugin/action/throttle/throttle.go]: done

File: plugin/action/throttle/redis_limiter.go (hunk 5 above)
Comment (HeadHunter483) [reply plugin/action/throttle/redis_limiter.go:207]: done

**Review by vadimalekseev — APPROVED** [review]

**Review by vano144 — APPROVED** [review]

**Pull request merged**